NTBS-184 add overview page with all notification info and placeholders for info yet to be added

ChristianMcc opened the pull request:
<!--- Provide a general summary of your changes in the Title above, including the related Jira ticket -->

## Description
<!--- High level changes overview -->

## Checklist:
- [X] Automated tests are passing locally.
### Accessibility testing
- [X] Test without javascript
- [X] Test in small window (immitating small screen)
- [X] Zoom page to 400% - content still visible?
- [X] Works with keyboard only
- [X] Tested with one screen reader
- [X] Passes automated checker (e.g. [WAVE](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh))

ChristianMcc commented on ntbs-service/Pages/Notifications/OverviewPartials/_ClinicalDetailsOverviewPartial.cshtml:
```diff
@@ -0,0 +1,98 @@
+    <div class="nhsuk-grid-row notification-overview-details">
+        <div class="nhsuk-grid-column-one-third">
+            <div class="notification-details-label"> BCG vaccination </div>
+            <div class="empty-cell-min-height"> @Model.Notification.ClinicalDetails.BCGVaccinationState @(@Model.Notification.ClinicalDetails.BCGVaccinationYear != null ? " - " + @Model.Notification.ClinicalDetails.BCGVaccinationYear : "") </div>
```
this logic will also be used later for the MDR treatment, probably treading the line of the longest amount of logic to be kept in a cshtml file, let me know what you think

jasiekmiko replied:
Ah! Well, the comment just above addresses this too, really :) But yes, agreed that it's creeping up on comfort limits :D

ChristianMcc commented on ntbs-service/Pages/Notifications/OverviewPartials/_ClinicalDetailsOverviewPartial.cshtml:
```diff
@@ -0,0 +1,98 @@
+    <div class="nhsuk-grid-row notification-overview-details">
+        <div class="nhsuk-grid-column-one-third">
+            <div class="notification-details-label"> Sites of Disease </div>
+            <div class="empty-cell-min-height"> @Model.SitesOfDisease </div>
```
empty-cell-min-height is used to create some space between fields when data is missing - especially useful while we have a lot of placeholders

jasiekmiko replied:
Sounds good - let's put a comment on the class definition in the css file.
Also, as its applied to all cells not just the empty ones, I'd suggest we don't use `empty`  in its name

ChristianMcc commented on ntbs-service/Pages/Notifications/OverviewPartials/_PatientOverviewPartial.cshtml:
```diff
@@ -0,0 +1,93 @@
+        <div class="nhsuk-grid-column-one-third">
+            <div class="notification-details-label"> NHS number </div>
+            <div class="empty-cell-min-height"> @(@Model.Notification.PatientDetails.NhsNumber.ToString().Substring(0, 3) + " " + @Model.Notification.PatientDetails.NhsNumber.ToString().Substring(3, 3) 
+                + " " + @Model.Notification.PatientDetails.NhsNumber.ToString().Substring(6, 4) ) </div>
```
Likewise this is pretty long, happy to move to cshtml.cs file

ffarmanov commented on ntbs-service/Pages/Notifications/OverviewPartials/_ClinicalDetailsOverviewPartial.cshtml:
```diff
@@ -0,0 +1,98 @@
+</div> 
+<div class="notification-overview-details-container">
+    <div class="nhsuk-grid-row notification-overview-details">
+        <div class="nhsuk-grid-column-one-third">
```
https://github.com/nhsuk/frontend-dotnetcore/tree/master/src/NHSUKFrontEndLibraryTagHelpers/NHSUK.FrontEndLibrary.TagHelpers/Tags/Layout

You can use template in this link. We should be using this library as long as we can for consistency and less boilerplate

jasiekmiko commented on ntbs-service/Pages/Notifications/Overview.cshtml:
```diff
@@ -0,0 +1,28 @@
+
+<partial name="./OverviewPartials/_PreviousHistoryOverviewPartial" model="@Model">
+
+<back-to-top-button inline-template>
```
As it's javascript-only, we don't want to be showing this button without javascript. We could use `v-if`, or not use an `inline-template` and instead provide the template in javascript (see `template` property on https://vuejs.org/v2/guide/components.html#Base-Example)
Alternatively, we could implement the scroll back up with a link to the top of the page ( `<a href='#top'>` where `top` is the id of an element), which is what https://design-system.service.gov.uk/components/ does, so perhaps nicest for us to follow?

jasiekmiko commented on ntbs-service/Pages/Notifications/Overview.cshtml.cs:
```diff
@@ -0,0 +1,82 @@
+            ImprisonmentRiskFactorTimePeriods = CreateTimePeriodsString(Notification.SocialRiskFactors.RiskFactorImprisonment);
+
+            var postcodeNoWhiteSpace = Notification.PatientDetails.Postcode.Replace(" ", string.Empty);
+            Postcode = postcodeNoWhiteSpace.Insert(postcodeNoWhiteSpace.Length - 3, " ");
```
Let's factor this out into a `formatPostcode` method too?

lbbethke replied:
Or could we just instead define an (unmapped) method "DisplayPostcode" on PatientDetails, so it can be called directly from the patientdetails model?

jasiekmiko commented on ntbs-service/Pages/Notifications/Overview.cshtml.cs:
```diff
@@ -0,0 +1,82 @@
+            var postcodeNoWhiteSpace = Notification.PatientDetails.Postcode.Replace(" ", string.Empty);
+            Postcode = postcodeNoWhiteSpace.Insert(postcodeNoWhiteSpace.Length - 3, " ");
+
+            for(var i = 0; i < Notification.NotificationSites.Count; i++) {
```
Same here, let's turn it into a method of its own.
Also, you might find it neater if written a bit more functionally, along the lines of:
```
var siteNames = Notification.NotificationSites.Select(ns => ns.Site).Select(s => s.Description)
SitesOfDisease = string.join(", ", siteNames)
```

lbbethke replied:
Again, I'd define this method directly on Notification, rather than in some kind of helper - what do you think?

jasiekmiko commented on ntbs-service/Pages/Notifications/Overview.cshtml.cs:
```diff
@@ -0,0 +1,82 @@
+            return Page();
+        }
+
+        public string CreateTimePeriodsString(RiskFactorBase riskFactor) {
```
Similarly here, I would suggest using [String.join](https://docs.microsoft.com/en-us/dotnet/api/system.string.join?view=netcore-2.2) to simplify this :)

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_ClinicalDetailsOverviewPartial.cshtml:
```diff
@@ -0,0 +1,98 @@
+
+<div class="notification-overview-type-and-edit-container">
+    <h3 class="notification-details-type"> Clinical details </h3>
+    <a href="./Edit/ClinicalDetails?id=@notificationId" class="notification-edit-link"> Edit </a>
```
Can I suggest that we also add a class to it that means it doesn't change colour when visited?
See https://design-system.service.gov.uk/styles/typography/#links, and maybe double check with Sorcha that's what we want :)

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_ClinicalDetailsOverviewPartial.cshtml:
```diff
@@ -0,0 +1,98 @@
+        </div>
+        <div class="nhsuk-grid-column-one-third">
+            <div class="notification-details-label"> Days from onset to treatment </div>
+            <div class="empty-cell-min-height"> @((@Model.Notification.ClinicalDetails.TreatmentStartDate?.Date - @Model.Notification.ClinicalDetails.SymptomStartDate?.Date)?.Days) </div>
```
Hmm, there's actually a bunch of these transformations around..
I wonder if we want to have a `DisplayNotification.cs` model, which would commonise all of these, as well as the ones performed in the controller atm - then it would be the only BoundProprerty on the page?

lbbethke replied:
That ties in with what I said above :) I definitely think some refactoring might be in order! There is the formatted date model too, which is reused in multiple places so probably should be in its own model

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_ClinicalDetailsOverviewPartial.cshtml:
```diff
@@ -0,0 +1,98 @@
+        </div>
+        <div class="nhsuk-grid-column-one-third">
+            <div class="notification-details-label"> Drug resistance profile </div>
+            <div>  </div>
```
class missing?

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_PatientOverviewPartial.cshtml:
```diff
@@ -0,0 +1,93 @@
+        </div>
+        <div class="nhsuk-grid-column-one-third">
+            <div class="notification-details-label"> Sex </div>
+            <div class="empty-cell-min-height"> @if(@Model.Notification.PatientDetails.Sex != null) {
```
will `... Sex?.Label` work?

jasiekmiko commented on ntbs-service/wwwroot/css/site.css:
```diff
@@ -118,6 +118,44 @@ body {
     width: 100px;
 }
 
+.notification-overview-type-and-edit-container {
```
Should we add another css file for the view page?

jasiekmiko commented on ntbs-service/wwwroot/source/app.ts:
```diff
@@ -12,13 +12,15 @@ import { DateComparison } from './Components/DateComparison';
 import { ValidateContactTracing } from './Components/ValidateContactTracing';
 import { ValidateRiskFactor } from './Components/ValidateRiskFactor';
-// Vue needs to be the firs thing to load!
+// Vue needs to be the first thing to load!
```
Thank you 😂

ChristianMcc commented:
Moved a lot of the logic to Notification.cs, added tag helpers and removed back to top javascript button and added id="top" to header of main layout (with the intention that this means any page can have the same functionality)

jasiekmiko commented on ntbs-service/Models/Notification.cs:
```diff
@@ -1,18 +1,127 @@
+        }
+
+        public int? CalculateDaysBetweenNullableDates(DateTime? date1, DateTime? date2) {
+            return (date1?.Date - date2?.Date)?.Days;
```
Not sure this works correctly..
If `date1` is null, then `date?.Date` is also null, at which point we'll be trying to subtract from null - causing an exception I imagine 🤔

ChristianMcc replied:
I've tested this and it seems to just return null if either date is missing and doesn't throw an exception, happy to check for nulls before running this to make it safer though?

jasiekmiko replied:
No, you're right, looks like c# is safer than i thought! :)

jasiekmiko commented on ntbs-service/Models/Notification.cs:
```diff
@@ -1,18 +1,127 @@
+            if(riskFactor.MoreThanFiveYearsAgo) {
+                timeString = String.Join(", ", new string[] {timeString, "more than 5 years ago"});
+            }
+            return timeString;
```
Sorry I don't think my comment about `String.Join` was super clear before. One of th enice things about it is it can take an array, so we can build up the array first and then concatenate it in one go. So something along the lines of:
```
            var timeStrings = new List<strings>();
            if(riskFactor.IsCurrent) {
                timeStrings.Add("current");
            }
            if(riskFactor.InPastFiveYears) {
                timeStrings.Add("less than 5 years ago");
            }
            if(riskFactor.MoreThanFiveYearsAgo) {
                timeStrings.Add("more than 5 years ago");
            }
           String.Join(", ", timeStrings);
```

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_ClinicalDetailsOverviewPartial.cshtml:
```diff
@@ -4,95 +4,95 @@
-            <div class="empty-cell-min-height"> @Model.Notification.ClinicalDetails.PresentationDate?.ToString("dd-MMM-yyyy") </div>
-        </div>
-        <div class="nhsuk-grid-column-one-third">
+            <div class="cell-min-height"> @Model.Notification.ClinicalDetails.PresentationDate?.ToString("dd-MMM-yyyy") </div>
```
I reckon the dates want to be display properties on the Notification model, too

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_ContactTracingOverviewPartial.cshtml:
```diff
@@ -4,47 +4,47 @@
             <div class="notification-details-label"> Number of contacts identified for screening </div>
             <div> Adults - @Model.Notification.ContactTracing.AdultsIdentified </div>
             <div> Children - @Model.Notification.ContactTracing.ChildrenIdentified </div>
             <div> Total - @(@Model.Notification.ContactTracing.AdultsIdentified + @Model.Notification.ContactTracing.ChildrenIdentified) </div>
```
Another potential candidate for property

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_PatientOverviewPartial.cshtml:
```diff
@@ -4,90 +4,87 @@
-             } </div>
-        </div>
-        <div class="nhsuk-grid-column-one-third">
+            <div class="cell-min-height"> @Model.Notification.PatientDetails.Sex?.Label </div>
```
Even these long chains might be better off there.. I'm sensing a trend here 😂

jasiekmiko commented on ntbs-service/Pages/Notifications/OverviewPartials/_PatientOverviewPartial.cshtml:
```diff
@@ -4,90 +4,87 @@
-            <div class="empty-cell-min-height"> @(@Model.Notification.PatientDetails.NoFixedAbode ? "No fixed abode" : @Model.Postcode) </div>
-        </div>
-        <div class="nhsuk-grid-column-one-third">
+            <div class="cell-min-height"> @(@Model.Notification.PatientDetails.NoFixedAbode ? "No fixed abode" : @Model.Notification.FormattedPostcode) </div>
```
...and another one :D

lbbethke commented on ntbs-service/Models/Notification.cs:
```diff
@@ -1,18 +1,127 @@
         public virtual ContactTracing ContactTracing { get; set; }
         public virtual SocialRiskFactors SocialRiskFactors { get; set; }
+
+        public string FullName {
```
I assume EF doesn't complain about these as they don't have setters, so it doesn't try to add them to the model?

jasiekmiko replied:
Oh good point!
This would suggest so: https://docs.microsoft.com/en-us/ef/core/modeling/included-properties#conventions
Worth checking though!
But if we needto exclude them, then: https://www.learnentityframeworkcore.com/configuration/fluent-api/ignore-method#excluding-a-property-from-mapping

lbbethke commented on ntbs-service/Models/Notification.cs:
```diff
@@ -1,18 +1,127 @@
+        }
+
+        public string DrugRiskFactorTimePeriods { 
+            get {
```
Wouldn't it make more sense for the methods that are defined on the owned models to be defined there, rather than all in here? Might we not want to make use of some of these  when we only have e.g. Clinical Details or PatientDetails available, so we don't have to go via notification? It would also make this model quite a bit shorter, and just add an extra ClinicalDetails into Notification.ClinicalDetails.DaysFromOnsetToTreament.
@jasiekmiko , @ChristianMcc , what do you think? Not sure whether you already had a chat in person

ChristianMcc replied:
Yep we've had a chat and I think it's an ongoing decision, I'll have another chat with Jan now to see if the current way is still preferred or to switch to the way you've described

jasiekmiko replied:
Yeah, my gut feeling is that we'll not be interacting with the individual tables much at all outside of the edit pages, and instead be manipulating the whole Notification, which means the current version is nicer, as the user does't have to care how the property is split up. Agreed that it's one to keep an eye on in the future and potentially shift over to individual pages if the need arises!
(Also, the `=>` syntax for getter-only properties should help with the lenght of the model, at least visually!)

lbbethke replied:
Ok, sounds good to me :)
